src/java.base/share/classes/java/lang/String.java

```diff
@@ -869,12 +882,10 @@ private String(Charset charset, byte[] bytes, int offset, int length) {
                 throw new Error(x);
             }
             if (COMPACT_STRINGS) {
-                byte[] bs = StringUTF16.compress(ca, 0, caLen);
-                if (bs != null) {
-                    value = bs;
-                    coder = LATIN1;
-                    return;
-                }
+                byte[] val = StringUTF16.compress(ca, 0, caLen);
+                this.coder = StringUTF16.coderFromArrayLen(val, caLen);
+                this.value = val;
+                return;
```

Comment:
This needs to call `initCompressionFlag()`:
```java
                if (coder == UTF16) {
                    initCompressionFlag();
                }
```

Reply:
Added